## List of apps and descriptions of them in this directory. 

* association_result_annotation: Annotate GWAS, PheWAS Assocaitions
  - association_result_annotation
* biobin: 
  - This app runs the latest development build of the rare variant binning tool BioBin.
* generate_phenotype_matrix: Generate Phenotype Matrix
  - generate_phenotype_matrix
* genotype_case_control: Generate Case/Control by Genotype
  - App provides case and control number by each genotype
* impute2: imputation
  - This will perfrom imputation using Impute2
* impute2_to_plink: Impute2 To PLINK
  - Convert Impute2 file to PLINK files
* plato_single_variant: PLATO - Single Variant Analysis 
  - Apps allows you to run single variant association testing against single phenotype (GWAS) or multiple phenotype (PheWAS) test
* rl_sleeper_app: sleeper
  - This App provides some useful tools when working with data in DNANexus. This App is designed to be run on the command line with "dx run --ssh RL_Sleeper_App" in the project that you have data that you want to explore (use "dx select" to switch projects as needed).
* shapeit2: SHAPEIT2
  - This app does phasing using SHAPEIT2
* strand_align: Strand Align
  - Strand Align prior to phasing
* vcf_annotation_formatter: 
  - Extracts and reformats VCF annotations (CLINVAR, dbNSFP, SIFT, SNPEff)
- **QC_apps folder:**
  * drop_marker_sample: Drop Markers and/or Samples (PLINK)
    * drop_marker_sample
  * drop_relateds: Relatedness Filter (IBD)
    * drop_relateds
  * extract_marker_sample: Drop Markers and/or Samples (PLINK)"
    * extract_marker_sample
  * maf_filter: Marker MAF Rate Filter (PLINK)
    * maf_filter
  * marker_call_filter: Marker Call Rate Filter (PLINK)
    * marker_call_filter
  * missing_summary: Missingness Summary (PLINK)
    * Returns missingness rate by sample
  * pca: Principal Component Analysis using SMARTPCA 
    * pca
  * sample_call_filter: Sample Call Rate Filter (PLINK)
    * sample_call_filter
